### ISSUE TITLE
torch.cuda.synchronize Influence distributed training

### ISSUE THREAD

**Issue by Richie-yan** [pytorch/pytorch#43947]
PyTorch :1.5.0
model：resnet50

Background : I ran distributed training of pytorch resnet50 in the same environment (using DataParallel to run a single machine with 4 cards), and ran two experiments respectively. The only difference between the two experiments was that `torch.cuda.synchronize()` was called after `loss.backward()`  , the other one is not called

Phenomenon: The training speed of calling synchronize is faster (0.345 s/step—> 0.276 s/step). Through nvprof, it is observed that there is a big difference in the time consumption of cudnn in the two experiments.

Doubt:
1. Why calling `torch.cuda.synchronize()` after `loss.backward()` will speed up the model training？
2. Why synchronize affect cudnn？
Hope to help answer, thanks.

**Comment by ezyang** [comment]
I don't know, this seems pretty mysterious. One possibility is you are selecting a different cudnn algorithm between the two runs; what if you set cudnn deterministic?

**Comment by Richie-yan** [comment]
I set torch.backends.cudnn.enabled = False
There is no speed difference between the two. It seems that torch.cuda.synchronize does affect some performance of cudnn. I need to look at this in detail.

**Comment by ptrblck** [comment]
Thank for reporting the issue @Richie-yan.
I'm can reproduce the performance increase with an additional synchronization using this code snippet:
```python
import torch
import torch.nn as nn
import torchvision.models as models
import time

torch.backends.cudnn.benchmark = True

model = models.resnet50()
model = nn.DataParallel(model, device_ids=[0, 1, 2, 3]).to('cuda:0')

x = torch.randn(64, 3, 224, 224, device='cuda:0')
target = torch.randint(0, 1000, (64,), device='cuda:0')
criterion = nn.CrossEntropyLoss()

# warmup
sync = False

out = model(x)
loss = criterion(out, target)
loss.backward()
if sync:
    torch.cuda.synchronize()

# fake training iters
for idx in range(100):
    if idx == 10:
        torch.cuda.synchronize()
        t0 = time.time()
    out = model(x)
    loss = criterion(out, target)
    loss.backward()
torch.cuda.synchronize()
t1 = time.time()
print((t1 - t0))
```
Could you please check it, if it would reflect your use case?

The results using `cudnn8.0.3.33`:
* no sync: 9.71555 +/- 0.05693
* with sync: 9.87690 +/- 0.11212

Since `cudnnFind` is profiling the kernels, the benchmarks seem to be affected negatively by potentially other workload on the device. While small profiling differences might be expected, it seems that `nn.DataParallel` is causing some additional overhead, which affects the end to end performance negatively. 

We'll check, if we could stabilize it.

**Comment by Richie-yan** [comment]
Hi, @ptrblck 
Yes, your test can reflect my case. 
I have another issue, which is also a weird phenomenon. Can you help us to take a look?
https://github.com/pytorch/pytorch/issues/44103
